packages/flutter_tools/lib/src/commands/build_aar.dart

```diff
@@ -0,0 +1,60 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import '../android/aar.dart';
+import '../build_info.dart';
+import '../project.dart';
+import '../runner/flutter_command.dart' show DevelopmentArtifact, FlutterCommandResult;
+import 'build.dart';
+
+class BuildAarCommand extends BuildSubCommand {
```

Comment:
According to https://codecov.io/gh/flutter/flutter/pull/35217/diff#D12-47, the runCommand method of this has no test coverage.

Reply:
I added unit testing to the methods that don't require leaking the entire implementation detail. 
While mocking is generally useful, it's not ideal to write unit tests that reveal how the underlaying code is implemented. Looking at the code, it seems that some of the issues can be mitigated by removing getters or special methods in classes like `FlutterProject`  that rely on files to exist at specific locations.


These methods are just wrapper around  The root cause is most getters like FlutterProject 

See https://news.ycombinator.com/item?id=7809402

Reply:
I added unit testing to the methods that don't require leaking the entire implementation detail.
While mocking is generally useful, it's not ideal to write unit tests that reveal how the underlaying code is implemented.  In other words, changing the implementation detail of a method shouldn't result in changes to the unit test.

Looking at the code, this seems to be caused by code expecting files at specific location. e.g. `FlutterProject` getters such as `pubspecFile`, `packagesFile` that are used downstream.

Relevant reads: https://news.ycombinator.com/item?id=7809402, https://medium.com/javascript-scene/mocking-is-a-code-smell-944a70c90a6a

Reply:
Some examples of how to test build commands exit in `test/general.shard/commands/build_*.dart`. This command, and buildAar has several areas where it is possible for regressions/changes to go unnoticed. I'm also expecting that the functionality here will grow over time, so even testing something trivial initially will make it easier for us down the line.

BuildAarCommand should correctly pass its arguments through the buildAar. We should ensure that defaults work as expected, and unsupported values aren't passed through uninitialized. This might be difficult to do in this case, because buildAar is just a top level function. Its a sort of unhelpful pattern that we're using across the build* commands, so it might be best to take a different approach with build AAR.

buildAar itself has several checks for project files/module/gradle plugin. We'll want to ensure those are triggered correctly, given that the underlying detection works. This could be done through mocking, or through the memory file system.

buildGradleAar does a lot of configuring. I'm not quite sure how to start here.

Reply:
I hear what you are saying. I think there are a few classes that could be refactored, so it's clear what the boundaries are as a result.  For example, a `FlutterProject` or `AndroidProject` shouldn't preserve state by depending on  `file_system`. There should be factories and methods that operate on those types instead.  This way it's easier to construct one of these types in a unit test.

Currently, if I want to create a `FlutterProject` module, I need to create a file with specific content to test that a method operates on a module project only 😕, but that is just one case. In practice, you end up with many mocks that aren't related to the code that is being tested. 

Also, it would be useful to aggregate code coverage from integration test into the final code coverage report. I'm sure I'm testing these paths there, but it would be good to get that insight in the report.  wdyt?

Reply:
Please don't hesitate to refactor stuff to make it testable. Embrace the yak shave. The entirety of my current (40+ file) PR is a result of my trying to do a tiny cleanup and just following all the threads that came out of that. It's normal for one thing to lead to another in this way. :-)

Reply:
Thanks for chiming in, @Hixie.  I filed https://github.com/flutter/flutter/issues/36424, and I'm more than happy to address this issue holistically. :)

Reply:
I will refactor the current code that calls Gradle to get the list of tasks, properties, and build right after this PR.